HSEARCH-3042 Allow the definition of one "search" analyzer per field as part of the mapping

fax4ever opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-3042

~~Preview because:~~
~~1. Still need a test using a `ANALYZER_NGRAM` ( common case for having a search analyzer )~~ done!
~~2. Try what happens if we try to override a normalizer instead...~~ done!
~~3. Still missing multi-indexes compatibility check tests~~ done!
~~4. Documentation~~ done!

yrodiere commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/types/dsl/impl/ElasticsearchStringIndexFieldTypeOptionsStep.java:
```diff
@@ -130,6 +129,7 @@ public ElasticsearchStringIndexFieldTypeOptionsStep aggregable(Aggregable aggreg
 		if ( analyzerName != null ) {
 			mapping.setType( DataTypes.TEXT );
 			mapping.setAnalyzer( analyzerName );
+			mapping.setSearchAnalyzer( searchAnalyzerName );
```
You should probably add a check in the other branch of the `if`, to throw an exception if we have a search analyzer without an analyzer.
Same for the Lucene backend, I think.

Also, this would deserve a test in `org.hibernate.search.integrationtest.backend.tck.document.DocumentModelDslIT`, in my opinion.

fax4ever replied:
I think the same

yrodiere commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/types/predicate/impl/ElasticsearchTextFieldPredicateBuilderFactory.java:
```diff
@@ -42,7 +44,9 @@ public boolean hasCompatibleAnalyzer(ElasticsearchFieldPredicateBuilderFactory o
 		ElasticsearchTextFieldPredicateBuilderFactory castedOther = (ElasticsearchTextFieldPredicateBuilderFactory) other;
-		return Objects.equals( analyzer, castedOther.analyzer ) && Objects.equals( normalizer, castedOther.normalizer );
+		return Objects.equals( analyzer, castedOther.analyzer ) &&
+				Objects.equals( searchAnalyzer, castedOther.searchAnalyzer ) &&
```
If the search analyzers are the same but the analyzers are different, this should return true, I think. And the tests should probably reflect that... ?
In fact, you could solve this easily by only keeping the search analyzer in this class, and using the analyzer as a default when the search analyzer is `null`.

fax4ever replied:
You're right!

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/MatchSearchPredicateIT.java:
```diff
@@ -781,18 +781,21 @@ public void analyzerOverride() {
 				.predicate( f -> f.match().field( whitespaceAnalyzedField ).matching( "NEW WORLD" ) )
 				.toQuery();
 
+		// Terms are never lower-cased, nether at write nor at query time.
```
Maybe these comments would be easier to understand if they were above the query they are referring to? I was confused for a minute. You could remove the blank line between the query and the `assertThat` to make it clearer that they are related.

fax4ever commented on mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/mapping/definition/annotation/FullTextField.java:
```diff
@@ -55,6 +55,15 @@
+	 * <p>
+	 * As above, see the documentation of your backend to know how to define analyzers.
+	 */
+	String searchAnalyzer() default "";
```
Right. I tested only at backends!

fax4ever commented on engine/src/main/java/org/hibernate/search/engine/backend/types/dsl/StringIndexFieldTypeOptionsStep.java:
```diff
@@ -28,6 +28,18 @@
+	/**
+	 * Overrides {@link #analyzer} to use for query parameters at search time.
+	 * <p>
+	 * Requires analyzed type.
```
Sounds good

fax4ever commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/document/model/esnative/impl/PropertyMapping.java:
```diff
@@ -58,6 +57,8 @@
 	/**
 	 * Must be null when we don't want to include it in JSON serialization.
+	 * <p>
+	 * Populated using Java reflection.
```
Right! It was accessed by `PropertyMappingJsonAdapterFactory#addFields(Builder)`. I removed the field from here too.

yrodiere replied:
It's still there. You apparently renamed it to add an `s`... ? Also the getter is still there and returns `null`.

fax4ever replied:
Sorry @yrodiere, I did mess with the rebase. :(
I fixed it, you will see that on the next push.

fax4ever commented:
@yrodiere thanks for the comments on the preview. I think I've addressed them and done the remaining tasks here. So it's ready for review. Thanks

fax4ever commented:
[Here](https://ci.hibernate.org/job/hibernate-search/view/change-requests/job/PR-2099/9/) is a build to test all supported ES versions.

yrodiere commented on integrationtest/backend/lucene/src/test/java/org/hibernate/search/integrationtest/backend/lucene/search/LuceneMatchSearchPredicateIT.java:
```diff
@@ -66,21 +69,51 @@ public void match_skipAnalysis_normalizedStringField() {
+			normalizedSearchAnalyzedField = root.field(
+					"normalizedSearchAnalyzedField",
+					c -> c.asString().normalizer( DefaultAnalysisDefinitions.NORMALIZER_LOWERCASE.name )
+							.searchAnalyzer( OverrideAnalysisDefinitions.ANALYZER_WHITESPACE_LOWERCASE.name )
```
If you want to allow a search analyzer overriding a normalizer, you should also:

* add the option to `@KeywordField`
* test it
* update the documentation of the `searchAnalyzer` method in `StringIndexFieldTypeOptionsStep` (it only mentions overriding the analyzer).

fax4ever replied:
I added the test to probe the backends.
So it seems that the Lucene backend potentially can support it, in contrast Elasticsearch not.

I we added this option to `@KeywordField`, shouldn't it be supported by both backends?

If you want we can talk on zulip.

fax4ever replied:
I'll drop this commit, as we decided on zulip

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/document/DocumentModelDslIT.java:
```diff
@@ -476,6 +476,32 @@ public void analyzerAndNormalizer() {
+	@Test
+	public void searchAnalyzerWithoutAnalyzer() {
+		SubTest.expectException(
+				"Setting a search analyzer, without setting a normalizer on the same field",
```
`without setting a normalizer` => `without setting an analyzer`.

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/MatchSearchPredicateIT.java:
```diff
@@ -781,28 +781,46 @@ public void analyzerOverride() {
 		String whitespaceLowercaseAnalyzedField = indexMapping.whitespaceLowercaseAnalyzedField.relativeFieldName;
+		String whitespaceLowercaseSearchAnalyzedField = indexMapping.whitespaceLowercaseSearchAnalyzedField.relativeFieldName;
 
+		// Terms are never lower-cased, nether at write nor at query time.
```
`nether` => `neither`

fax4ever replied:
done x8 times :)

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/MatchSearchPredicateIT.java:
```diff
@@ -979,22 +979,32 @@ public void analyzerOverride_normalizedStringField() {
+		// we expect the same result.
+		query = scope.query()
+				.predicate( f -> f.match().field( ngramFieldPath ).matching( "worldofwordcraft" ) )
+				.toQuery();
```
Great idea! This is indeed very relevant.

fax4ever replied:
Thanks!

yrodiere commented on integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/MatchSearchPredicateIT.java:
```diff
@@ -56,6 +56,7 @@
 	private static final String INCOMPATIBLE_INDEX_NAME = "IndexWithIncompatibleFields";
 	private static final String INCOMPATIBLE_ANALYZER_INDEX_NAME = "IndexWithIncompatibleAnalyzer";
+	private static final String COMPATIBLE_SEARCH_ANALYZER_INDEX_NAME = "IndexWithCompatibleSearchAnalyzer";
 	private static final String INCOMPATIBLE_DECIMAL_SCALE_INDEX_NAME = "IndexWithIncompatibleDecimalScale";
```
Commit message:

> Test analyzer search multi-indexescompatibility

Maybe you meant: 

> Test search analyzer multi-index compatibility

?

yrodiere commented on documentation/src/main/asciidoc/mapper-orm-mapping.asciidoc:
```diff
@@ -263,6 +263,8 @@ match fields ignoring diacritics,
 Full-text fields must be assigned an <<mapper-orm-directfieldmapping-analyzer,analyzer>>, referenced by its name.
 See <<concepts-analysis>> for more details about analyzers and full-text analysis.
+Moreover, you can define <<mapper-orm-directfieldmapping-search-analyzer,a specific analyzer>>
+to interpret the query parameters at query time.
```
"a search analyzer to analyze searched terms differently"?

yrodiere commented on documentation/src/main/asciidoc/mapper-orm-mapping.asciidoc:
```diff
@@ -371,6 +373,13 @@ Only available on `@FullTextField`.
 
+[[mapper-orm-directfieldmapping-search-analyzer]] `searchAnalyzer`::
+An optional different analyzer, overriding the one defined with the `analyzer` attribute,
+to use only for query parameters at search time.
```
"to use only when analyzing searched terms"?

fax4ever commented:
Thanks @yrodiere. I think I've addressed the comments.
So I'll merge it as soon as [this full CI job](https://ci.hibernate.org/job/hibernate-search/view/change-requests/job/PR-2099/12/) works.

fax4ever commented:
Merged thanks!
